### DISCUSSION TITLE
"Severe recall & mAP drop after peak (~30 epochs) on custom 2-class dataset with YOLO26x – overfitting or hyperparams?"

### DISCUSSION THREAD

**Discussion by YooSeungHo0124** [ultralytics/ultralytics#23871] [Q&A]
**Hi Ultralytics team & community!**

I'm fine-tuning **YOLO26x** on a custom 2-class safety belt detection dataset and seeing a classic **peak-then-crash** behavior on validation metrics.

### Dataset
- ~150K training images (90/10 train/val split)
- Classes: 0=safety_belt (64.4%), 1=no_safety_belt (35.6%)
- imgsz=640, objects are mostly medium/large (people wearing/not wearing belts)

### Training setup (main args)
```yaml
epochs: 200
batch: 48 (RTX 3090 x4)
optimizer: AdamW
lr0: 0.001    # scaled for batch 48
lrf: 0.1
cos_lr: True
warmup_epochs: 5
weight_decay: 0.0005
box: 7.5
cls: 0.75     # boosted for minority class
dfl: 1.5

# Aug
hsv_h/s/v: 0.04/0.8/0.6
degrees: 10.0
translate: 0.1
scale: 0.5
fliplr: 0.5
mosaic: 1.0
mixup: 0.1
copy_paste: 0.1
close_mosaic: 10
```

<img width="1267" height="647" alt="tensorboard" src="https://github.com/user-attachments/assets/5b60b753-2c86-4c13-8db5-ab3ca624db0c" />

Metrics peak around epoch 25–35:
mAP50: ~0.94
mAP50-95: ~0.57–0.58
Precision: ~0.94–0.95
Recall: ~0.90

After that, sharp decline (especially recall):
At epoch 50: mAP50 ≈0.776, mAP50-95 ≈0.467, Recall ≈0.69, Precision ≈0.88

Training loss keeps decreasing slowly, but val metrics collapse → classic overfitting sign?
Recall drop is the most dramatic → model becomes very conservative (misses many positives, low FN explosion)

### Questions
Is this expected / common with YOLO26x (NMS-free, MuSGD, etc.) on custom datasets?
Any recommended hyperparameter tweaks to prevent late-stage collapse? (e.g. lower cls gain, reduce late aug, change scheduler, lower wd, etc.)
Should I early-stop at peak and resume from best.pt with much lower lr (e.g. 1e-5 ~ 3e-5)?
Any YOLO26-specific tips (e.g. close_mosaic timing, ProgLoss interaction, small-object handling)?

Thanks in advance for any advice — really appreciate the community's help!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @YooSeungHo0124, thank you for the detailed report and for using Ultralytics 🚀! This is an automated response to help get your discussion into the best shape for review, and an Ultralytics engineer will also assist soon 😊

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users and advanced workflows, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since your post includes a possible training/validation behavior issue, if this may be a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For training-related discussions like this, it also helps to include:
- The exact training command or Python script used
- Your `data.yaml`
- The exact `ultralytics` package version
- A few dataset image examples and label samples
- Full training logs, validation logs, and any relevant plots
- Whether the behavior reproduces after upgrading and retraining on the latest package

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks for already sharing the metric plot and training settings 📊—that context is very helpful. If you can add the items above, especially a minimal reproducible example and exact version/command details, the team will be in a much better position to review what’s happening with your YOLO26x training run.

**Reply by glenn-jocher** [reply]
`cls=0.75` is a global classification-loss gain, so it won’t specifically reweight your minority class; this kind of late recall/mAP drop is usually overfitting or train/val augmentation mismatch rather than something expected from YOLO26x itself, so I’d keep `best.pt`, enable early stopping with `patience`, and retrain from the same base weights with lighter late-stage aug (`mixup=0`, `copy_paste=0`, lower `mosaic`, earlier `close_mosaic`) before trying a very low-LR resume, then use the PR/F1 and confusion outputs in the [model evaluation guide](https://docs.ultralytics.com/guides/model-evaluation-insights/) and [metrics deep dive](https://docs.ultralytics.com/guides/yolo-performance-metrics/) to see whether one class is collapsing; if you can attach `results.csv` and `confusion_matrix_normalized.png`, we can narrow it down further.

**Comment by Sir-Sloth-The-Lazy** [comment]
This is classic overfitting on a small dataset. With ~150K training images the model
is large enough to memorize rather than generalize. A few things to try:

1. **Early stopping** — add `patience=20` to your train() call. YOLO will stop when
   val metrics stop improving.

2. **Stronger augmentation** — set `mosaic=1.0`, `mixup=0.3`, `copy_paste=0.1` in
   your train() args. These help a lot on 2-class datasets where the model sees limited
   variety.

3. **Regularization** — bump `weight_decay` from 0.0005 to 0.001, add `dropout=0.1`.

4. **Smaller model** — YOLO26x is massive. For a 2-class task, YOLO26m or YOLO26l
   usually generalizes better unless your classes have extreme scale variation.

5. **Learning rate schedule** — use `cos_lr=True` with `lrf=0.01`. A cosine decay
   prevents the optimizer from driving weights too far on later epochs.

If recall is crashing while precision stays high, your confidence threshold may be
drifting, also check your val set for class imbalance.

**Reply by glenn-jocher** [reply]
150K train images is not a small dataset, so I’d treat this less as “tiny-data overfitting” and more as an overly custom long-run recipe: for a 200-epoch run `optimizer=auto` would normally switch to `MuSGD`, so I’d first remove manual `AdamW`/`lr0`, go back closer to default fine-tuning settings, add `patience=20`, and note that `cls` is a global loss gain rather than minority-class weighting, so use `cls_pw` if imbalance is the goal; the [fine-tuning guide](https://docs.ultralytics.com/guides/finetuning-guide/) and [YOLO26 training recipe](https://docs.ultralytics.com/guides/yolo26-training-recipe/) cover the same pattern, and yes, keep `best.pt` from the peak run before trying any lower-LR second stage or a smaller `YOLO26l`/`YOLO26m` baseline.
